Discogs import lists

aglowinthefield opened the pull request:
#### Database Migration
NO

#### Description
A few sentences describing the overall goals of the pull request's commits.

#### Screenshot (if UI related)

#### Todos
- [x] Tests
- [x] Translation Keys (./src/NzbDrone.Core/Localization/Core/en.json)
- [x] [Wiki Updates](https://wiki.servarr.com)

#### Issues Fixed or Closed by this PR

* Fixes #3629

aglowinthefield committed:
Discogs list by ID

aglowinthefield committed:
Discogs list by ID

aglowinthefield committed:
Wantlists

aglowinthefield committed:
Wantlists

RetroDadson commented on .gitignore:
This shouldn't be committed as part of this change

RetroDadson commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsWantlistParser.cs:
```diff
@@ -0,0 +1,140 @@
+                catch
+                {
+                    // If we can't fetch release details, skip this item
+                    continue;
```
This should log what has failed as people will ask "why isn't x in" and there is no clear understanding of it from this

aglowinthefield replied:
https://github.com/Lidarr/Lidarr/pull/5577/commits/9792c298d92ba6bb202463c1ab0590770720b6b7

RetroDadson commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsListsParser.cs:
```diff
@@ -0,0 +1,148 @@
+{
+    public string Name { get; set; }
+    public int Id { get; set; }
+}
```
These should be defined in separate files similar to Spotify

RobinDadswell reviewed (changes requested):
A few changes needed, also this will need squashing into a single commit with the title "New: Discog Import List support" or something similar

mynameisbogdan commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsListsParser.cs:
```diff
@@ -0,0 +1,148 @@
+                catch
+                {
+                    // If we can't fetch release details, skip this item
+                    continue;
```
`continue` is redundant, it's better to be replaced with a `logger.Error`.

mynameisbogdan commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsLists.cs:
```diff
@@ -0,0 +1,38 @@
+        public override ImportListType ListType => ImportListType.Discogs;
+        public override TimeSpan MinRefreshInterval => TimeSpan.FromHours(12);
+        public override TimeSpan RateLimit => TimeSpan.FromSeconds(3); // Conservative rate limiting to avoid 429 errors when fetching many releases
+        public override int PageSize => 0; // Discogs doesn't support pagination for lists
```
```suggestion
        public override TimeSpan RateLimit => TimeSpan.FromSeconds(3);
```

mynameisbogdan commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsLists.cs:
```diff
@@ -0,0 +1,38 @@
+
+        public override IImportListRequestGenerator GetRequestGenerator()
+        {
+            return new DiscogsListsRequestGenerator { Settings = Settings };
+        }
+
+        public override IParseImportListResponse GetParser()
+        {
+            var parser = new DiscogsListsParser();
+            parser.SetContext(_httpClient, Settings);
+            return parser;
+        }
```
```suggestion
        public override IImportListRequestGenerator GetRequestGenerator()
        {
            return new DiscogsListsRequestGenerator(Settings);
        }

        public override IParseImportListResponse GetParser()
        {
            return new DiscogsListsParser(Settings, _httpClient, _logger);
        }
```

mynameisbogdan commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsListsParser.cs:
```diff
@@ -0,0 +1,148 @@
+        }
+
+        return items;
+    }
```
```suggestion
    private readonly DiscogsListsSettings _settings;
    private readonly IHttpClient _httpClient;
    private readonly ILogger _logger;

    public DiscogsListsParser(DiscogsListsSettings settings, IHttpClient httpClient, Logger logger)
    {
        _settings = settings;
        _httpClient = httpClient;
        _logger = logger;
    }

    public IList<ImportListItemInfo> ParseResponse(ImportListResponse importListResponse)
    {
        var items = new List<ImportListItemInfo>();

        if (!PreProcess(importListResponse))
        {
            return items;
        }

        var jsonResponse = Json.Deserialize<DiscogsListResponse>(importListResponse.Content);

        if (jsonResponse?.Items == null)
        {
            return items;
        }

        foreach (var item in jsonResponse.Items)
        {
            if (item.Type == "release" && item.ResourceUrl.IsNotNullOrWhiteSpace())
            {
                try
                {
                    var releaseInfo = FetchReleaseDetails(item.ResourceUrl);

                    items.AddIfNotNull(releaseInfo);
                }
                catch (Exception ex)
                {
                    _logger.Error(ex, "Discogs release details API call resulted in an unexpected exception");
                }
            }
        }

        return items;
    }
```

mynameisbogdan commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsListsParser.cs:
```diff
@@ -0,0 +1,148 @@
+
+        foreach (var item in jsonResponse.Items)
+        {
+            if (item?.Type == "release" && item.ResourceUrl.IsNotNullOrWhiteSpace())
```
Seems like this only supports `release` type items. Is that correct? In their [API](https://www.discogs.com/developers#page:user-lists,header:user-lists-list) I noticed a type for `artist` as well.

aglowinthefield replied:
It does support `artist` but I was under the impression lidarr lists are meant to resolve to a list of release groups, effectively. So I filtered them out. Is that wrong @mynameisbogdan ?

mynameisbogdan replied:
Lists support both artists and albums (release groups), meaning if someone only has artists it might not import anything. 

Plainly put, if the list item only has the artist title and no album name, it's going to import the whole artist.

aglowinthefield replied:
Okay, so sounds like I should support both for the sake of keeping things consistent w/ other list behaviors then?

mynameisbogdan replied:
Yeah, that's my take too.

aglowinthefield replied:
Think my latest change should cover it, added some tests too to be sure. Assumed album could just be null https://github.com/Lidarr/Lidarr/pull/5577/commits/d900890e6ba033e73151a64cf56ee0b22d18fc5a

aglowinthefield replied:
I was tempted to just use the displayName returned but figured it'd be better to circrle back to the API and get the true artist name to be safe.

mynameisbogdan commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsListsRequestGenerator.cs:
```diff
@@ -0,0 +1,36 @@
+                .Build();
+
+            yield return new ImportListRequest(request);
+        }
```
```suggestion
        private readonly DiscogsListsSettings _settings;

        public DiscogsListsRequestGenerator(DiscogsListsSettings settings)
        {
            _settings = settings;
        }

        public virtual ImportListPageableRequestChain GetListItems()
        {
            var pageableRequests = new ImportListPageableRequestChain();
            pageableRequests.Add(GetPagedRequests());

            return pageableRequests;
        }

        private IEnumerable<ImportListRequest> GetPagedRequests()
        {
            var request = new HttpRequestBuilder(_settings.BaseUrl.TrimEnd('/'))
                .Resource($"/lists/{_settings.ListId}")
                .SetHeader("Authorization", $"Discogs token={_settings.Token}")
                .Build();

            yield return new ImportListRequest(request);
        }
```

mynameisbogdan commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsListsSettings.cs:
```diff
@@ -0,0 +1,38 @@
+
+    public class DiscogsListsSettings : IImportListSettings
+    {
+        private static readonly DiscogsListsSettingsValidator Validator = new DiscogsListsSettingsValidator();
```
```suggestion
        private static readonly DiscogsListsSettingsValidator Validator = new ();
```

mynameisbogdan commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsWantlist.cs:
```diff
@@ -0,0 +1,38 @@
+        public override ImportListType ListType => ImportListType.Discogs;
+        public override TimeSpan MinRefreshInterval => TimeSpan.FromHours(12);
+        public override TimeSpan RateLimit => TimeSpan.FromSeconds(3); // Conservative rate limiting to avoid 429 errors when fetching many releases
+        public override int PageSize => 0; // Discogs doesn't support pagination for wantlists currently
```
[They do according to their API](https://www.discogs.com/developers#page:user-wantlist,header:user-wantlist-wantlist):
> Returns the list of releases in a user’s wantlist. Accepts Pagination parameters.

aglowinthefield replied:
added!

mynameisbogdan commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsWantlist.cs:
```diff
@@ -0,0 +1,38 @@
+using System;
```
Same suggestions from `DiscogsLists*` apply to `DiscogsWantlist*` too.

mynameisbogdan commented on src/NzbDrone.Core/ImportLists/Discogs/DiscogsWantlistParser.cs:
```diff
@@ -0,0 +1,140 @@
+                {
+                    if (_httpClient != null && _settings != null)
+                    {
+                        var releaseInfo = FetchReleaseDetails(want.BasicInformation.ResourceUrl);
```
If you want is `artists.First().Name` and `title`, then fetching the release details is redundant according to their API.

https://www.discogs.com/developers#page:user-wantlist,header:user-wantlist-wantlist


```json
{
     "basic_information": {
        "formats": [
          {
            "text": "Digipak",
            "qty": "1",
            "descriptions": [
              "Album"
            ],
            "name": "CD"
          }
        ],
        "thumb": "https://api-img.discogs.com/PsLAcp_I0-EPPkSBaHx2t7dmXTg=/fit-in/150x150/filters:strip_icc():format(jpeg):mode_rgb()/discogs-images/R-1867708-1248886216.jpeg.jpg",
        "cover_image": "https://api-img.discogs.com/PsLAcp_I0-EPPkSBaHx2t7dmXTg=/fit-in/500x500/filters:strip_icc():format(jpeg):mode_rgb()/discogs-images/R-1867708-1248886216.jpeg.jpg",
        "title": "Year Zero",
        "labels": [
          {
            "resource_url": "https://api.discogs.com/labels/2311",
            "entity_type": "",
            "catno": "B0008764-02",
            "id": 2311,
            "name": "Interscope Records"
          }
        ],
        "year": 2007,
        "artists": [
          {
            "join": "",
            "name": "Nine Inch Nails",
            "anv": "",
            "tracks": "",
            "role": "",
            "resource_url": "https://api.discogs.com/artists/3857",
            "id": 3857
          }
        ],
        "resource_url": "https://api.discogs.com/releases/1867708",
        "id": 1867708
      },
      "resource_url": "https://api.discogs.com/users/example/wants/1867708",
      "id": 1867708
    }
```

aglowinthefield committed:
Clean up and don't fetch release details

aglowinthefield committed:
Clean up and don't fetch release details

aglowinthefield committed:
Undo extra gitignore

aglowinthefield committed:
Undo extra gitignore

aglowinthefield committed:
Add logging on list parser errors

aglowinthefield committed:
Add logging on list parser errors

aglowinthefield committed:
Update src/NzbDrone.Core/ImportLists/Discogs/DiscogsLists.cs

Co-authored-by: Bogdan <mynameisbogdan@users.noreply.github.com>

aglowinthefield committed:
Update src/NzbDrone.Core/ImportLists/Discogs/DiscogsLists.cs

Co-authored-by: Bogdan <mynameisbogdan@users.noreply.github.com>

aglowinthefield committed:
Update src/NzbDrone.Core/ImportLists/Discogs/DiscogsLists.cs

Co-authored-by: Bogdan <mynameisbogdan@users.noreply.github.com>

aglowinthefield committed:
Update src/NzbDrone.Core/ImportLists/Discogs/DiscogsLists.cs

Co-authored-by: Bogdan <mynameisbogdan@users.noreply.github.com>

aglowinthefield committed:
Update src/NzbDrone.Core/ImportLists/Discogs/DiscogsListsRequestGenerator.cs

Co-authored-by: Bogdan <mynameisbogdan@users.noreply.github.com>

aglowinthefield committed:
Update src/NzbDrone.Core/ImportLists/Discogs/DiscogsListsRequestGenerator.cs

Co-authored-by: Bogdan <mynameisbogdan@users.noreply.github.com>

aglowinthefield committed:
move API DTOs to separate file

aglowinthefield committed:
move API DTOs to separate file

aglowinthefield committed:
Merge branch 'feat/discogs-import-list' of https://github.com/aglowinthefield/Lidarr into feat/discogs-import-list

aglowinthefield committed:
Update src/NzbDrone.Core/ImportLists/Discogs/DiscogsListsSettings.cs

Co-authored-by: Bogdan <mynameisbogdan@users.noreply.github.com>

aglowinthefield committed:
Update src/NzbDrone.Core/ImportLists/Discogs/DiscogsListsSettings.cs

Co-authored-by: Bogdan <mynameisbogdan@users.noreply.github.com>

aglowinthefield committed:
More cleanup and DI

aglowinthefield committed:
More cleanup and DI

aglowinthefield committed:
Support pagination

aglowinthefield committed:
Support pagination

aglowinthefield commented:
@mynameisbogdan @RobinDadswell Addressed feedback and did some major cleanup after getting brushed up on C# a bit. Should answer everything except one outstanding question about `artist` vs `release` types.

aglowinthefield committed:
Support artist type from discogs api

aglowinthefield committed:
Support artist type from discogs api

aglowinthefield committed:
Stylecop fixes

aglowinthefield committed:
Stylecop fixes
